### DISCUSSION TITLE
Hiding columns using a view

### DISCUSSION THREAD

**Discussion by yepes** [supabase/supabase#17692] [Questions]
Hi,

I'm trying to exclude fields to be returned on the public api. 

My approach is to have a private schema not accesible to the api and the views with a subset of fields in the public one. I'm not exactly sure if that's the correct approach.

What I have so far:

```
create schema privateapi;

create table "privateapi"."testtable" (
"id" bigint generated by default as identity not null,
"created_at" timestamp with time zone not null default now(),
"public_field" text not null,
"private_field" text
);

alter table "privateapi"."testtable" enable row level security;

CREATE UNIQUE INDEX testtable_pkey ON privateapi.testtable USING btree (id);
alter table "privateapi"."testtable" add constraint "testtable_pkey" PRIMARY KEY using index "testtable_pkey";
create or replace view "public"."testapi" as SELECT testtable.id,
testtable.created_at,
testtable.public_field
FROM privateapi.testtable;

create policy "Enable read access for all users"
on "privateapi"."testtable"
as permissive
for select
to authenticated
using (true);

insert into privateapi.testtable (public_field) values('public 1');
insert into privateapi.testtable (public_field) values('public 2');
insert into privateapi.testtable (public_field) values('public 3');
```

With the above code, no rls are taking place. If I just: 
```
curl --location 'http://localhost:54321/rest/v1/testapi'
```
The three rows are returned.

So my (near to none) experience with postgres tells me that the query is executed by the user of supabase who created the table, so no rls.

Then, I changed the view to security_invoker:
```
alter view public.testapi set (security_invoker = true);
```

I logged in as an user, queried the ep with the access_token:

```
curl --location 'http://localhost:54321/rest/v1/testapi' \
--header 'Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJhdXRoZW50aWNhdGVkIiwiZXhwIjoxNjk1NjI5NzAxLCJpYXQiOjE2OTU2MjYxMDEsInN1YiI6ImM5NmMzMTJlLTU0MjMtNDQxMi04YzJlLTBlYTllMmYxNGJhMyIsImVtYWlsIjoiamVzdXNnb2x0cmF0ZWMtMTIzQGdtYWlsLmNvbSIsInBob25lIjoiIiwiYXBwX21ldGFkYXRhIjp7InByb3ZpZGVyIjoiZW1haWwiLCJwcm92aWRlcnMiOlsiZW1haWwiXSwidGlkIjoiYWIzMThjYWUtOWFiZC00N2E4LTg2MDctZjQ4N2NmN2E1Y2Y0In0sInVzZXJfbWV0YWRhdGEiOnt9LCJyb2xlIjoiYXV0aGVudGljYXRlZCIsImFhbCI6ImFhbDEiLCJhbXIiOlt7Im1ldGhvZCI6InBhc3N3b3JkIiwidGltZXN0YW1wIjoxNjk1NjI2MTAxfV0sInNlc3Npb25faWQiOiJjYTU3YmUzMy02NTk2LTRiMDUtOWUxOC1kOTlmNTc2YjI1YWYifQ.V0Tdas4clqV-njHZV7gjmvT03H14nupwQ9lcxodS4nU'
```

Which returns:
```
{"code":"42501","details":null,"hint":null,"message":"permission denied for table testtable"}%
```

I'm a bit lost from here. Any tips about how to resolve the error of about the correct approach (if mine is wrong) would be appreciated.

Thanks

**Comment by GaryAustin1** [comment]
Well with invoker you are no longer running as postgres user so don't have access to the schema and tables in the custom schema as you did not grant auth/authenticated/service_role access.
https://supabase.com/docs/guides/api/using-custom-schemas

**Reply by yepes** [reply]
Can you provide an example? 

If I just create the schema, table, view, enlable RLS and apply the privileges to the schema, a simple curl will return all the data:

```sql
create schema if not exists "privateapi";

create table "privateapi"."testtable" (
    "id" bigint generated by default as identity not null,
    "created_at" timestamp with time zone not null default now(),
    "public_field" text not null,
    "private_field" text,
    "tenant_id" uuid
);


create or replace view "public"."testapi" as  SELECT testtable.id,
    testtable.created_at,
    testtable.public_field
   FROM privateapi.testtable;

alter table "privateapi"."testtable" enable row level security;

GRANT USAGE ON SCHEMA privateapi TO anon, authenticated, service_role;
GRANT ALL ON ALL TABLES IN SCHEMA privateapi TO anon, authenticated, service_role;
GRANT ALL ON ALL ROUTINES IN SCHEMA privateapi TO anon, authenticated, service_role;
GRANT ALL ON ALL SEQUENCES IN SCHEMA privateapi TO anon, authenticated, service_role;
ALTER DEFAULT PRIVILEGES FOR ROLE postgres IN SCHEMA privateapi GRANT ALL ON TABLES TO anon, authenticated, service_role;
ALTER DEFAULT PRIVILEGES FOR ROLE postgres IN SCHEMA privateapi GRANT ALL ON ROUTINES TO anon, authenticated, service_role;
ALTER DEFAULT PRIVILEGES FOR ROLE postgres IN SCHEMA privateapi GRANT ALL ON SEQUENCES TO anon, authenticated, service_role;
```

If RLS is enabled and nothing is applied, shouldn't I have to get an empty array (or an error) as a response?

**Reply by GaryAustin1** [reply]
I don't have any examples I can recall of someone showing doing a view on an alternate schema.
What happened to `alter view public.testapi set (security_invoker = true);`
Without that it will ignore RLS.

Also you might as in Discord.

**Reply by yepes** [reply]
I think alter view public.testapi set (security_invoker = true); was the key. I think I have it working, thanks!

**Comment by EdgarsLv** [comment]
this is what documentation says:
ALTER VIEW public.blocked_users SET (security_invoker = on)

22023: unrecognized parameter "security_invoker". 
What i'm doing wrong?

**Reply by ericmillsio** [reply]
It's only in Postgres 15, you'd have to upgrade your DB. I'm not sure what the alternative solution is for this in 14.
